### ISSUE TITLE
Poor performance with larger code base

### ISSUE THREAD

**Issue by davidgiga1993** [Guardsquare/proguard#283]
Hi,

I'm using proguard for nearly 10 years now and so far I never had any major performance issues. However at some point proguard got extremely slow - I can't recall the exact time and I assume it's related to the bytecode input. It takes about 25 minutes to obfuscate a java desktop application with about 400kLoc on an Mac Mini M1.
I've also tested it on an Windows 11 Ryzen 5950x and it took 22 minutes. Most of this time only 1 core is used, sometimes 2 for a short period.

The project is currently using `com.guardsquare:proguard-gradle:7.2.2` with gradle 7.5 and temurin jdk 17 with 1.8 as compile target.

The app uses quite a bit of autogenerated code with huge switch/case statements, and I somehow suspect that it could be related to that, howver I haven't verified it yet.

My questions are:
1. Is there a way to profile what exactly takes proguard so long?
2. Is performance something you want improve in the future?

Happy to get any feedback on this topic. In the meantime I'll try to remove the auto generated code and see if it makes a difference.

For reference, here is the proguard config and output of the M1 build:
```
-optimizations !code/simplification/cast,!field/*,!class/merging/*
-optimizationpasses 3
-allowaccessmodification

-adaptresourcefilecontents **.properties,META-INF/MANIFEST.MF

# The remainder of this file is identical to the non-optimized version
# of the Proguard configuration file (except that the other file has
# flags to turn off optimization).

-dontusemixedcaseclassnames
-dontskipnonpubliclibraryclasses
-verbose

-keepattributes *Annotation*

-keepclassmembers enum * {
    public static **[] values();
    public static ** valueOf(java.lang.String);
}

-keepclassmembers class **.R$* {
    public static <fields>;
}

# Libgdx lwjgl fix
-keep class org.lwjgl.system.** { *; }

# External libraries
-dontwarn org.xmlpull.v1.**
-dontwarn org.slf4j.**

# Jetty
-keep class org.eclipse.** { *; }
-keepclasseswithmembers class * {
@org.eclipse.** *;
}
-keepclassmembers class * {
@org.eclipse.** *;
}
-dontwarn "org.eclipse.jetty.**"
```

```
Explaining why classes and class members are being kept...
Printing usage to [/Users/david/jenkins/workspace/appName/desktop/build/usage.txt]...
Removing unused program classes and class elements...
  Original number of program classes:            11757
  Final number of program classes:               9962
Optimizing (pass 1/3)...
  Number of finalized classes:                   4797
  Number of unboxed enum classes:                0
  Number of vertically merged classes:           0   (disabled)
  Number of horizontally merged classes:         0   (disabled)
  Number of merged wrapper classes:              0   (disabled)
  Number of removed write-only fields:           0   (disabled)
  Number of privatized fields:                   0   (disabled)
  Number of generalized field accesses:          0   (disabled)
  Number of specialized field types:             0   (disabled)
  Number of inlined constant fields:             0   (disabled)
  Number of privatized methods:                  2512
  Number of staticized methods:                  877
  Number of finalized methods:                   21555
  Number of desynchronized methods:              0
  Number of simplified method signatures:        911
  Number of removed method parameters:           1156
  Number of generalized method invocations:      0
  Number of specialized method parameter types:  0
  Number of specialized method return types:     0
  Number of inlined constant parameters:         664
  Number of inlined constant return values:      78
  Number of inlined short method calls:          4674
  Number of inlined unique method calls:         3197
  Number of inlined tail recursion calls:        44
  Number of merged code blocks:                  104
  Number of variable peephole optimizations:     32078
  Number of arithmetic peephole optimizations:   1509
  Number of cast peephole optimizations:         0   (disabled)
  Number of field peephole optimizations:        137
  Number of branch peephole optimizations:       5602
  Number of object peephole optimizations:       2377
  Number of string peephole optimizations:       1822
  Number of math peephole optimizations:         0
  Number of simplified instructions:             2046
  Number of removed instructions:                24610
  Number of removed local variables:             1275
  Number of removed exception blocks:            611
  Number of optimized local variable frames:     9113
Shrinking...
Removing unused program classes and class elements...
  Original number of program classes:            9962
  Final number of program classes:               9736
Optimizing (pass 2/3)...
  Number of finalized classes:                   4
  Number of unboxed enum classes:                0
  Number of vertically merged classes:           0   (disabled)
  Number of horizontally merged classes:         0   (disabled)
  Number of merged wrapper classes:              0   (disabled)
  Number of removed write-only fields:           0   (disabled)
  Number of privatized fields:                   0   (disabled)
  Number of generalized field accesses:          0   (disabled)
  Number of specialized field types:             0   (disabled)
  Number of inlined constant fields:             0   (disabled)
  Number of privatized methods:                  10
  Number of staticized methods:                  104
  Number of finalized methods:                   116
  Number of desynchronized methods:              0
  Number of simplified method signatures:        384
  Number of removed method parameters:           508
  Number of generalized method invocations:      0
  Number of specialized method parameter types:  0
  Number of specialized method return types:     0
  Number of inlined constant parameters:         834
  Number of inlined constant return values:      41
  Number of inlined short method calls:          470
  Number of inlined unique method calls:         193
  Number of inlined tail recursion calls:        1
  Number of merged code blocks:                  3
  Number of variable peephole optimizations:     2895
  Number of arithmetic peephole optimizations:   52
  Number of cast peephole optimizations:         0   (disabled)
  Number of field peephole optimizations:        0
  Number of branch peephole optimizations:       174
  Number of object peephole optimizations:       0
  Number of string peephole optimizations:       430
  Number of math peephole optimizations:         0
  Number of simplified instructions:             975
  Number of removed instructions:                18026
  Number of removed local variables:             707
  Number of removed exception blocks:            0
  Number of optimized local variable frames:     678
Shrinking...
Removing unused program classes and class elements...
  Original number of program classes:            9736
  Final number of program classes:               9732
Optimizing (pass 3/3)...
  Number of finalized classes:                   1
  Number of unboxed enum classes:                0
  Number of vertically merged classes:           0   (disabled)
  Number of horizontally merged classes:         0   (disabled)
  Number of merged wrapper classes:              0   (disabled)
  Number of removed write-only fields:           0   (disabled)
  Number of privatized fields:                   0   (disabled)
  Number of generalized field accesses:          0   (disabled)
  Number of specialized field types:             0   (disabled)
  Number of inlined constant fields:             0   (disabled)
  Number of privatized methods:                  0
  Number of staticized methods:                  44
  Number of finalized methods:                   23
  Number of desynchronized methods:              0
  Number of simplified method signatures:        108
  Number of removed method parameters:           132
  Number of generalized method invocations:      0
  Number of specialized method parameter types:  0
  Number of specialized method return types:     0
  Number of inlined constant parameters:         404
  Number of inlined constant return values:      32
  Number of inlined short method calls:          0
  Number of inlined unique method calls:         7
  Number of inlined tail recursion calls:        0
  Number of merged code blocks:                  2
  Number of variable peephole optimizations:     244
  Number of arithmetic peephole optimizations:   2
  Number of cast peephole optimizations:         0   (disabled)
  Number of field peephole optimizations:        0
  Number of branch peephole optimizations:       7
  Number of object peephole optimizations:       43
  Number of string peephole optimizations:       12
  Number of math peephole optimizations:         0
  Number of simplified instructions:             159
  Number of removed instructions:                4741
  Number of removed local variables:             127
  Number of removed exception blocks:            0
  Number of optimized local variable frames:     43
Shrinking...
Removing unused program classes and class elements...
  Original number of program classes:            9732
  Final number of program classes:               9732
Obfuscating...
Printing mapping to [/Users/david/jenkins/workspace/appName/desktop/build/mapping.txt]...
  Number of obfuscated classes:                  6306
  Number of obfuscated fields:                   18156
  Number of obfuscated methods:                  22881
Preverifying...
Writing output...
```

**Comment by mrjameshamilton** [comment]
Hi @davidgiga1993 !

Great to hear you're a long time user of ProGuard!

There are a couple of issues reported with performance reduction already (#79 #91) but there was never any conclusion or reproducible sample.

> Is there a way to profile what exactly takes proguard so long?

An easy way to narrow down the timing of each feature is to pass the log output through the `ts` command to add a timestamp to every line of logging output. That way you can see which parts take longest.

```
java -jar proguard.jar @config.pro | ts '[%Y-%m-%d %H:%M:%S]' &> build.log
```

> Is performance something you want improve in the future?

We don't have any specific plans but if you can provide some reproducible sample and we can narrow down the issue that will be a good start for solving any performance problems.

**Comment by davidgiga1993** [comment]
Thanks for the fast reply!
Not sure how helpful the output will be since most of the time is spent after printing `Optimizing (pass X/3)...`. Each pass takes about  6-8 minutes. I even reduced the number of passes, in the past I was using 5 but it just took too long.

Similar to the other issues, providing a sample will be a bit tricky since it's the IP of my company, but since each optimization step takes so long, would it be possible to (privately) provide the application jar after one pass of obfuscation?

Also as a sidenode, when building the app for android, R8 takes about 2 minutes with the same proguard config file.

```
(I'll put the full timestamped log here once it's done)
```

**Comment by davidgiga1993** [comment]
I did a bit more profiling and it looks like the most time is spent in `proguard.evaluation.Variables.generalize(proguard.evaluation.Variables, boolean)`.
Drilling down a bit 44% of that methods time is spent in `proguard.evaluation.value.InstructionOffsetValue.contains(int)` which is a simple loop search. Replacing it with an HashMap for fast lookups might be a solution.

![image](https://user-images.githubusercontent.com/3924958/194698083-8def1aee-5908-4d66-a17f-5d114cc57eaf.png)

**Comment by mrjameshamilton** [comment]
Nice work tracking that down!

**Comment by EricLafortune** [comment]
I originally gave InstructionOffsetValue an array of ints instead of a HashSet of Integers because the array typically only contains one or two elements and there are many instances of them during the 
abstract evaluation of a method. So my goal was to keep the memory use and garbage collection in check. Automatically generated code with huge switch/case statements breaks the assumption. You should 
perhaps check the increase in memory use for large methods.

**Comment by mrjameshamilton** [comment]
Hi @davidgiga1993 ! Thanks for your contribution. The fix you made is now in ProGuard 7.3.1.
